### PR TITLE
refactor(table): Upgrade @availity/table to Typescript and Add Table Controls

### PR DESCRIPTION
**Pull request by smashleez**
This pull request appears to be more than it is. This includes upgrading the Table to Typescript.

The new feature being added is Table Controls :
1) BulkTableActions -A dropdown that will select/deselect all records and perform actions on them.
2) TableSorter - A dropdown that allows you to change the sort field and direction.

Also to create a Provider to house the table instance so that more than just the Table itself knows the context.

Currently, the storybook story is just completely busted - I just want to get these typescript conversions and table controls into an actual version outside of alpha and will circle back and investigate more.

### CODE REVIEW

File: docusaurus/docs/components/table/tableControls.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+---
+title: <TableControls/>
+---
+
+`<TableControls/>` is a container that wraps any given components that would be utilized in concert with the table component. This includes bulk selection, bulk actions, pagination, or any custom component that could potentially be needed. There are two components provided that can be utilized within the `<TableControls>`: `<BulkTableActions/>` and `<TableSorter/>`.
+
+### Props
+
+#### `id?: string`
+This is a unique id that used to identity the component.
```
Comment (smashleez) [inline docusaurus/docs/components/table/tableControls.md]: identify

File: docusaurus/docs/components/table/tableControls.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+);
+```
+
+## `<BulkTableActions/>`
+This component will allow you to select all and deselect all records in the table and perform actions on them. 
+
+### Props
+
+#### `id?: string`
+This is a unique id that used to identity the component.
```
Comment (smashleez) [inline docusaurus/docs/components/table/tableControls.md]: identify

File: docusaurus/docs/components/table/tableControls.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+This is the action that is taken whenever the action is selected from the dropdown, passing the records that are currently selected in the table.
+
+`isVisible?: (selectedRecords: object[]) => void`
+This is a function that determines whether the given action should display, passing the records that are currently selected in the table.
+
+## `<TableSorter/>`
+This components supplies a dropdown of all the available sort fields, alongside a toggle that will sort that field ascending or descending. This dropdown is automatically populated by what columns allow sorting when set in the `<TableProvider>`.
+
+#### `id?: string`
+This is a unique id that used to identity the component.
```
Comment (smashleez) [inline docusaurus/docs/components/table/tableControls.md]: identify

File: docusaurus/docs/components/table/tableControls.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+#### `disabled?: boolean`
+If this flag is set to true, the dropdown and buttons will be disabled.
+
+#### `color?: string` 
+This identifies the color variant to use for the dropdown buttons. For what variants are available, refer to Reactstrap.
+#### `onSort?: (sortBy: TableSort) => void`
+
+Event handler that is called when data is sorted. This passes in current sort configuration for the table.
```
Comment (smashleez) [inline docusaurus/docs/components/table/tableControls.md]: passes in the

File: packages/table/src/CellDefinitions/ActionCell.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import Icon from '@availity/icon';
+import TableActionMenu from '../TableActionMenu';
+import TableActionMenuItem from '../TableActionMenuItem';
+import { Cell, IdType } from '../types/ReactTable';
+import { RecordAction } from '../types/RecordAction';
+import { PrimaryRecordAction } from '../types/PrimaryRecordAction';
+
+export interface ActionCellConfig<T> {
+  actions: RecordAction<T>[];
+  primaryAction?: PrimaryRecordAction<T>
```
Comment (smashleez) [inline packages/table/src/CellDefinitions/ActionCell.tsx]: semicolon

File: packages/table/src/Controls/TableSorter.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import { TableSortOption } from '../types/TableSortOption';
+
+type Props = {
+    id?: string;
+    disabled?: boolean;
+    color?: string;
+    onSort?: (sortBy: TableSort[]) => void;
+} & React.HTMLAttributes<HTMLElement>;
+
+/* eslint-disable-next-line  @typescript-eslint/ban-types */
```
Comment (smashleez) [inline packages/table/src/Controls/TableSorter.tsx]: remove disable

File: packages/table/src/types/ReactTable.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
+import {
```
Comment (smashleez) [inline packages/table/src/types/ReactTable.ts:1]: The types are not correct from React-Table, so the go to recommendation from the Google is to create our own extensions of them. This is what this file is doing.

File: packages/table/styles.scss [hunk 8]
```diff
@@ -227,3 +227,9 @@ thead > tr {
 		padding: 5px !important;
 	}
 }
+
+.table-controls {
+	.table-controls-sorting {
+		margin-left: 15px;
```
Comment (smashleez) [inline packages/table/styles.scss:233]: Maybe can just use bootstrap classes?

File: rollup.config.js [hunk 9]
```diff
@@ -34,7 +34,8 @@ export default () =>
         commonjs(),
         typescript({
           tsconfigOverride: {
-            files: [`${pkgPath}/src/index.ts`],
```
Comment (smashleez) [inline rollup.config.js:37]: If you want to package types, just including the index.ts file is not enough.

**Comment by smashleez** [comment]
@jordan-a-young Any way you would be able to review/approve this? Trying to get this out there. Mostly the changes are converting the table to typescript and adding table actions.

**Comment by smashleez** [comment]
I keep having to backmerge master into this. If we can get this reviewed/approved/merged in that would be appreciated!

**Comment by jordan-a-young** [comment]
> I keep having to backmerge master into this. If we can get this reviewed/approved/merged in that would be appreciated!

I'm sorry about taking a while on this. I'll make sure to check it out today.

File: packages/table/src/CellDefinitions/ActionCell.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  primaryAction?: PrimaryRecordAction<T>;
+}
+
+const ActionCell = <T extends IdType>({
+  actions,
+  primaryAction,
+}: ActionCellConfig<T>): ((cell: Cell<T>) => JSX.Element) => {
+  const ActionCellDef = ({ row: { original }, index }: Cell<T>): JSX.Element => (
+    <>
+      <TableActionMenu id={`table_row_action_menu_${original.id ? original.id : index}`}>
```
Comment (jordan-a-young) [inline packages/table/src/CellDefinitions/ActionCell.tsx:20]: We can shorten this to `original.id || index`

File: packages/table/src/Controls/TableControls.tsx [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import { Form } from 'reactstrap';
+
+type Props = {
+  id?: string;
+  disabled?: boolean;
+  children?: React.ReactNode[] | React.ReactNode;
+} & React.HTMLAttributes<HTMLElement>;
+
+const TableControls = ({ id, disabled, children, ...rest }: Props): JSX.Element => {
+  const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);
```
Comment (jordan-a-young) [inline packages/table/src/Controls/TableControls.tsx:11]: It's ok to leave out the `<boolean>` here. It will get inferred from `disabled`.

File: packages/table/src/Controls/TableSorter.tsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+  disabled,
+  color,
+  sortOptions,
+  autoGenerateSortOptions,
+  ...rest
+}: Props): JSX.Element | null => {
+  const { sortableColumns, instance } = useTableContext();
+  const { toggleSortBy, state } = instance as TableInstance<T>;
+
+  const [isSortingDropdownOpen, setIsSortingDropdownOpen] = useState<boolean>(false);
```
Comment (jordan-a-young) [inline packages/table/src/Controls/TableSorter.tsx:41]: We can removeÏ each of these `useState`s that have `<boolean>`. The others should stay since they are initialized to `undefined`

`<TableSortOption | undefined>` can become `<TableSortOption>`. The `undefined` will be inferred from the initial value

File: packages/table/src/TableProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  columns,
+  data,
+  selectable,
+  scrollable,
+  sortable,
+  children,
+  ...rest
+}: TableProviderProps<T>): JSX.Element => {
+  let selectionColumn: Column<T>;
+  const [isScrollable, setScrollable] = useState<boolean | undefined>(scrollable);
```
Comment (jordan-a-young) [inline packages/table/src/TableProvider.tsx:35]: Same note about inferred types

**Review by jordan-a-young — DISMISSED** [review]

**Comment by jordan-a-young** [comment]
@smashleez I approved since most of my comments were nitpicks, and this is an alpha. I do wonder if we can come up with a design where the user won't have to import the `TableProvider`, but we can go over that as we move forward.

I saw you made a comment about storybook not working. It worked for me when I pulled the branch down, just make sure to have the `dist` built before running

**Review by jordan-a-young — APPROVED** [review]
